### ISSUE TITLE
[cssom] Shortest serialization principle with multiple minimal possibilities

### ISSUE THREAD

**Issue by Loirooriol** [w3c/csswg-drafts#8155]
https://drafts.csswg.org/cssom/#serialize-a-css-value

> If component values can be omitted or replaced with a shorter representation without changing the meaning of the value, omit/replace them.

But there are 2 different minimal ways to represent `list-style: outside none disc`:
 - `list-style: outside`
 - `list-style: disc`

And there are 3 different minimal ways to represent `border: medium none currentcolor`:
 - `border: medium`
 - `border: none`
 - `border: currentcolor`

And there are 4 different minimal ways to represent `text-decoration: none solid currentcolor auto`:
 - `text-decoration: none`
 - `text-decoration: solid`
 - `text-decoration: currentcolor`
 - `text-decoration: auto`

And so on with lots of shorthands. Which minimal serialization should be picked?

```js
var {style} = document.createElement("div");
style.listStyle = "outside none disc";
style.listStyle; // Firefox: "outside". Blink/WebKit: "outside none disc"
style.border = "medium none currentcolor";
style.border; // Firefox: "medium none". Blink/WebKit: "medium none currentcolor"
style.textDecoration = "none solid currentcolor auto";
style.textDecoration; // Firefox: "none". Blink: "none solid currentcolor". WebKit: "" (CSS3 syntax not implemented)
```

**Comment by tabatkins** [comment]
I propose as a general principle whenever it's a question of "which value do we need to provide, since they're *all* default", we choose the first one in grammar order. This matches with the existing principle that you serialize in grammar order. We can probably explicitly define this in reference to the grammar's top level being a || combinator (or a repetition of such), and say that anything still ambiguous is a spec bug that needs to be clarified.

This happens to match Firefox's behavior for 'list-style' and 'text-decoration'. Not sure why they serialize 'border' with the default thickness keyword *and* the default style keyword. Just to be less confusing, maybe? My proposal would have it serialize as just `"medium"`.

**Comment by cdoublev** [comment]
> whenever it's a question of "which value do we need to provide, since they're _all_ default", we choose the first one in grammar order

You may already know this but it has a [precedent](https://www.w3.org/TR/2011/WD-cssom-20110712/#serializing-css-values):

> Where CSS component values of the value can be omitted without changing the meaning of the value (e.g. initial values in shorthand properties), omit them. If this would remove all the values, then include the first allowed value.

**Comment by nt1m** [comment]
Don't we usually use the canonical order in this case?

**Comment by Loirooriol** [comment]
I think I haven't seen any property where the canonical order isn't "per grammar". So in practice it's the same?

**Comment by SebastianZ** [comment]
#13343 indicates that it might be confusing for authors if the serialization returns a different value as the one set. So, might it be possible to base the serialized value on the authored value somehow?

Still following the shortest serialization principle, the definition could be to use the first authored longhand value and skip all the rest of the longhand values that match their initial value.
That would result in serializing `border: currentcolor` to `currentcolor`, `border: none` to `none`, and `border: medium none currentcolor` to `medium`.

Sebastian

**Comment by emilio** [comment]
> So, might it be possible to base the serialized value on the authored value somehow?

You mean for shorthands in particular? In general I'd rather not, since it'd be on a case-by-case basis (on some properties the syntax might not allow the reordering or what not), and serialization is already under-specified enough :)

**Comment by SebastianZ** [comment]
> > So, might it be possible to base the serialized value on the authored value somehow?
> 
> You mean for shorthands in particular? In general I'd rather not, since it'd be on a case-by-case basis (on some properties the syntax might not allow the reordering or what not)

I mainly had shorthand serialization in mind, yes. Though I don't see why it couldn't apply to longhands, as well. The idea is to properly extend the algorithm to make it precise enough on when the rule [outlined earlier](https://github.com/w3c/csswg-drafts/issues/8155#issuecomment-3748877165) is applied.
Or in other words, the algorithm should provide a basic definition for how to serialize those cases. Other specifications may still provide different serialization rules if needed.

> ... and serialization is already under-specified enough :)

Well, that's the reason for issues like this one and #13343.😉 The main goal here is to improve the specification of the serialization and reduce ambiguities.

Sebastian

**Comment by Loirooriol** [comment]
I disagree with #13343 being a problem. CSS performs all kind of normalizations, so authors should just not expect to get the same exact thing they assigned, just an equivalent one.

**Comment by SebastianZ** [comment]
It surely isn't a big problem, but it is definitely something unexpected. And I believe we should consider the priority of constituencies here and try to meet author expectations as far as possible.

Sebastian

**Comment by Psychpsyo** [comment]
This might be more of a spec wording issue, but I'd like to throw it in anyways:
The spec says
> If component values can be omitted or replaced with a shorter representation without changing the meaning of the value, omit/replace them.

"shorter representation", to me, sounds like if there's a choice between two keywords, the shorter one wins (i.e. shorter in terms of actual resulting string length). Although this would be problematic for backwards compat because anytime a new, shorter keyword gets added, it'd need to be made an exception.

**Comment by fantasai** [comment]
Agenda+ When multiple components are omittable, but you need at least one of them, use the first one in the grammar, unless otherwise specified or otherwise required for compat.
